Add gamepad article (#4154)
GoogleChrome/web.dev: src/site/content/en/blog/gamepad/index.md

  - capabilities
  - gamepad
---

Chrome's offline page easter egg is one of the worst-kept secrets in history
(`[citation needed]`, but claim made for the dramatic effect).
If you press the <kbd>space</kbd> key or, on mobile devices, tap the dinosaur,
the offline page becomes an actually playable arcade game.
You might be aware of the fact that you do not actually have to go offline
when you feel like playing: You can also just navigate to `chrome://dino`, or,
for the geek in you, browse to
`chrome://network-error/-106`.
But did you know that there are currently
[270 million games played every month](https://www.blog.google/products/chrome/chrome-dino#jump-content:~:text=There%20are%20currently%20270%20million%20games%20played%20every%20month)?

<figure class="w-figure">
  <img class="w-screenshot w-screenshot--filled"
       src="dino.png"
       alt="Chrome's offline page with the Chrome dino game.">
  <figcaption class="w-figcaption">
    Press the space bar to play!
  </figcaption>
</figure>

Another fact that arguably is more useful to know and that you might not be aware of
is that the game can be played with a gamepad.

> Gamepad support was added exactly one year ago as of the time of this writing in a

Reviewer: Maybe mention that gamepad only works in arcade mode. (in other words if the URL is `chrome://dino`)

Author: Will do.

[commit](https://github.com/chromium/chromium/commit/fcafd36b23c535e307da4213b7d639f8c13b8da2)
by [Reilly Grant](https://github.com/reillyeon).
As you can see, the game, just like the rest of the Chromium project,
is fully
[open source](https://github.com/chromium/chromium/tree/master/components/neterror/resources).
In this post, I want to show you how to use the Gamepad API, and as a bonus feature,

> let you play the Chrome dino game on a Nintendo Switch (if you own one).

Reviewer: Where is this explained?

Author: You managed to review an article that was in flux ;-)

Author: @beaufortfrancois It's ready now :-)

    connected: true
    id: "Xbox 360 Controller (STANDARD GAMEPAD Vendor: 045e Product: 028e)"
    index: 0
    mapping: "standard"
    timestamp: 6563054.284999998
    vibrationActuator: GamepadHapticActuator {type: "dual-rumble"}
  */
});
```

### Being notified when a gamepad gets disconnected

Being notified of gamepad disconnects happens analogously to the way connections are detected.
Just this time the app listens for the `gamepaddisconnect` event.
Note how in the example below `connected` is now `false`.

```js
window.addEventListener('gamepadconnected', (event) => {
  console.log('❌ 🎮 A gamepad was disconnected:', event.gamepad);
  /*
    gamepad: Gamepad
    axes: (4) [0, 0, 0, 0]
    buttons: (17) [GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton]

>     connected: false

Reviewer: Could we use some emphasis on this line with mardown tools such as `js/6`?
I'm not sure about those numbers, you'll have to try it out ;)

Author: Sounds good!

    id: "Xbox 360 Controller (STANDARD GAMEPAD Vendor: 045e Product: 028e)"
    index: 0
    mapping: "standard"
    timestamp: 6563054.284999998
    vibrationActuator: null
  */
});
```

### The gamepad in your game loop

Getting a hold of a gamepad starts with a call to `navigator.getGamepads()`,
which returns a `GamepadList` object with `Gamepad` items.
Note that for this to work, you may need to "wake" your gamepad by pressing any of its buttons.
More than one gamepad can be connected to a device,
so be sure to check all items of the `GamepadList`.

```js
const pollGamepad = () => {
  // Always call `navigator.getGamepads()` inside of
  // the game loop, not outside.
  const gamepads = navigator.getGamepads();
  for (const gamepad of gamepads) {

>     if (!gamepad) {

Reviewer: Can you explain this line?
How could we get a falsy `gamepad` object from `gamepads`?

Author: ```js
pads = navigator.getGamepads()
// GamepadList {0: null, 1: null, 2: null, 3: null, length: 4}
for (const pad of pads) { 
  console.log(pad)
}
/*
VM221:2 null
VM221:2 null
VM221:2 null
VM221:2 null
*/

Reviewer: Is that a feature?

Reviewer: Oh!
https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/modules/gamepad/navigator_gamepad.h;l=97;drc=5b8933e94139a0ab5be46141666fdfcce0f624f6
```
  // A reference to the buffer containing the last-received gamepad state. May
  // be nullptr if no data has been received yet.
```

Author: Also gamepads remember their slot, a gamepad may not always be present at the first "free" slot.

Reviewer: You may want to add a comment for both then inline.

Author: Done.

      continue;
    }
    console.log(gamepad);
    // Process the gamepad state.
  }
  // Call yourself upon the next animation frame.
  // (Typically this happens every 60 times per second.)
  window.requestAnimationFrame(pollGamepad)
};
// Kick off the initial game loop iteration.

> window.requestAnimationFrame(pollGamepad)

Reviewer: It's not needed to call requestAnimationFrame here.
```suggestion
pollGamepad();
```

1. On the right-hand side in the **About** section, find the link to
   [https://tomayac.github.io/chrome-dino-gamepad/](https://tomayac.github.io/chrome-dino-gamepad/).
   Enjoy!
1. 🚨 For regular Switch online services to work again,
   turn your DNS settings back to **Automatic**.
   Conveniently, the Switch remembers previous manual DNS settings,
   so you can easily toggle between **Automatic** and **Manual**.

For the [Chrome dino gamepad](https://tomayac.github.io/chrome-dino-gamepad/) demo to work,
I have ripped out the Chrome dino game from the core Chromium project
(updating an [earlier effort](https://github.com/arnellebalane/trex-runner) by
[Arnelle Ballane](https://arnellebalane.com/)),
placed it on a standalone site, extended the existing gamepad API implementation by adding ducking
and vibration effects, created a full screen mode,
and [Mehul Satardekar](https://github.com/mehulsatardekar)
contributed a dark mode implementation.
Happy gaming!

## Useful links

- [Gamepad API spec](https://w3c.github.io/gamepad/)
- [Gamepad API extensions spec](https://w3c.github.io/gamepad/extensions.html)
- [GitHub repository](https://github.com/w3c/gamepad/)

## Acknowledgements

> This article was reviewed by [François Beaufort](https://github.com/beaufortfrancois).

Author: ```suggestion
This article was reviewed by [François Beaufort](https://github.com/beaufortfrancois)
and [Joe Medley](https://github.com/jpmedley).
```

---
title: Play the Chrome dino game with your gamepad
subhead: Learn how to use the Gamepad API to push your web games to the next level.
description: |
Learn to control web games with the Gamepad API.
authors:
  - thomassteiner

> date: 2020-10-30

Author: ```suggestion
date: 2020-11-03
# updated: 2020-11-03
```

---
title: Play the Chrome dino game with your gamepad

> subhead: Learn how to use the Gamepad API to push your web games to the next level.

Author: Yeah, I'm tired of clichés, too, but in this case it really does fit.

description: |
  This article teaches you how to use the Gamepad API to control web games.
authors:
  - thomassteiner
date: 2020-10-30
hero: hero.jpg

> alt: The hands of a person playing the Chrome Dino game on a Nintendo Switch.

Author: It's not. I'll change to "dino".

> alt: The hands of a person playing the Chrome Dino game on a Nintendo Switch.

Author: ```suggestion
alt: The hands of a person playing the Chrome dino game on a Nintendo Switch.
```

is fully
[open source](https://github.com/chromium/chromium/tree/master/components/neterror/resources).
In this post, I want to show you how to use the Gamepad API, and, as a bonus feature,
let you play the Chrome dino game on a Nintendo Switch (if you own one).

## Using the Gamepad API

{% Aside %}
  The [Gamepad API](https://w3c.github.io/gamepad/) has been around for a long time.
  This post disregards all the legacy features and vendor prefixes.
{% endAside %}

### Feature detection and browser support

The Gamepad API has universally great [browser support](https://caniuse.com/gamepad)
across both desktop and mobile.
You can detect if the Gamepad API is supported by running the feature test from the snippet below:

```js
if ('getGamepads' in navigator) {
  // The API is supported!
}
```

### How the browser represents a gamepad

The browser represents gamepads as `Gamepad` objects.
A `Gamepad` has the following fields:

- `id`:
  An identification string for the gamepad.
  This string identifies the brand or style of connected gamepad device.
- `index`:
  The index of the gamepad in the navigator.
- `connected`:

>   Indicates whether the gamepad is still connected to the system.

Author: Agreed, thanks!

- `timestamp`: The last time the data for this gamepad was updated.

> - `mapping`: The mapping in use for this device. Currently the only mapping is `"standard"`.

Author: Mapping in this case refers to the buttons and axes. An Xbox controller may have an A/B/X/Y button layout and a PlayStation controller may use B/A/X/Y or something.

> - `mapping`: The mapping in use for this device. Currently the only mapping is `"standard"`.

Author: ```suggestion
- `mapping`: The button and axes mapping in use for this device. Currently the only mapping is `"standard"`.
```

> - `mapping`: The mapping in use for this device. Currently the only mapping is `"standard"`.

Author: I made clearer what the mapping is about.

gamepads "remember" their slot and may not always be present at the first available slot.

```js
// When no gamepads are connected:
navigator.getGamepads()
// GamepadList {0: null, 1: null, 2: null, 3: null, length: 4}
```

If one or several gamepads are connected,
but `navigator.getGamepads()` still reports `null` items,
you may need to "wake" each gamepad by pressing any of its buttons.
You can then poll the gamepad states in your game loop as shown below.

```js
const pollGamepad = () => {
  // Always call `navigator.getGamepads()` inside of
  // the game loop, not outside.
  const gamepads = navigator.getGamepads();
  for (const gamepad of gamepads) {
    // Disregard empty slots.
    if (!gamepad) {
      continue;
    }
    // Process the gamepad state.
    console.log(gamepad);
  }
  // Call yourself upon the next animation frame.
  // (Typically this happens every 60 times per second.)
  window.requestAnimationFrame(pollGamepad)
};
// Kick off the initial game loop iteration.
pollGamepad();
```

{% Aside 'gotchas' %}
  Do not store a lasting reference to the `GamepadList` result *outside* of the game loop,
  since the method returns a static snapshot, not a live object.

>   Call `navigator.getGamepads()` each time anew in your game loop.

Author: No, actually this has to happen in the game loop. I'll add an emphasis.

>   Call `navigator.getGamepads()` each time anew in your game loop.

Author: ```suggestion
  Call `navigator.getGamepads()` each time anew *in your game loop*.
```

{% endAside %}

### Making use of the vibration actuator

> The `vibrationActuator` field corresponds to a configuration of motors or other actuators
> that can apply a force for the purposes of haptic feedback.

Reviewer 2: ```suggestion
The `vibrationActuator` property returns a `GamepadHapticActuator`
object, which corresponds to a configuration of motors or other actuators
that can apply a force for the purposes of haptic feedback.
```

Author: Great job digging this up. This is the way it happens in Canary, not yet Beta. But for future-proofness I think it's fine to say how it will work. Thank you.

> Effects can be played by calling `Gamepad.vibrationActuator.playEffect()`.

Reviewer 2: haptic effects?

Author: Indeed.

> Effects can be played by calling `Gamepad.vibrationActuator.playEffect()`.

Author: ```suggestion
Haptic effects can be played by calling `Gamepad.vibrationActuator.playEffect()`.
```

### Integration with Permissions Policy

The Gamepad API spec defines a
[policy-controlled feature](https://w3c.github.io/webappsec-permissions-policy/)
identified by the string `"gamepad"`. Its default `allowlist` is `"self"`.
A document's permissions policy determines whether any content in that document is allowed
to access `navigator.getGamepads()`.
If disabled in any document, no content in the document will be allowed to use
`navigator.getGamepads()`, nor will the `gamepadconnected` and
`gamepaddisconnected` events fire.

```html/2
<iframe
    src="index.html"
    allow="gamepad"
>
</iframe>
```

## Demo

A simple [gamepad tester demo](https://gamepad-demo.glitch.me/) is embedded below.
The source code is available [on Glitch](https://glitch.com/edit/#!/gamepad-demo).
Try the demo by connecting a gamepad via USB or Bluetooth and pressing any of its buttons
or moving any of its axis.

<div class="glitch-embed-wrap" style="height: 1000px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/gamepad-demo?path=index.html&previewSize=100"
    title="gamepad-demo on Glitch"
    allow="gamepad"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

> ## Bonus: Play Chrome dino on a Nintendo Switch

Reviewer 2: ```suggestion
## Bonus: play Chrome dino on a Nintendo Switch
```

The Nintendo Switch contains a
[hidden browser](https://switchbrew.org/wiki/Internet_Browser#WifiWebAuthApplet),

> which essentially just serves for logging in to Wi-Fi networks behind a captive portal.

Reviewer 2: ```suggestion
which serves for logging in to Wi-Fi networks behind a captive portal.
```

The browser is pretty barebones and does not have a URL bar,

> but, once you have navigated to a page, it is fully usable.

Reviewer 2: How do you navigate to a page without a URL bar?

Author: With Google :-) This comes a little below.

When doing a connection test in system settings, the Switch will detect that the captive portal
is present and display an error for it when the response for
[http://conntest.nintendowifi.net/](http://conntest.nintendowifi.net/)
does not include the `X-Organization: Nintendo` HTTP header.

> We can make creative use of this by pointing the Switch to a DNS server

Reviewer 2: ```suggestion
I can make creative use of this by pointing the Switch to a DNS server
```

> that simulates to be a captive portal that then redirects to a search engine.

Reviewer 2: ```suggestion
that simulates a captive portal that then redirects to a search engine.
```

1. Go to **System Settings** and then **Internet Settings** and
   find the Wi-Fi network that your Switch is connected to. Tap **Change Settings**.
1. Find the section with the **DNS Settings** and add
   [45.55.142.122](http://45.55.142.122) as a new **Primary DNS**.
   Note that this DNS server is *not operated by Google*
   but a [third-party](https://www.switchbru.com/dns/), so proceed at your own risk.

> 1. **Save** the settings and then tap **Connect to This Network**.
> 1. The Switch will tell you that **Registration is required to use this network**. Tap **Next**.

Reviewer 2: ```suggestion
1. **Save** the settings and then tap **Connect to This Network**. The Switch will tell you that **Registration is required to use this network**. 
1. Tap **Next**.
```

Reviewer 2: Steps always begin with a user action. Responses are always at the end of an action.

Author: Good call.

> 1. On the page that opens, make your way through to **Google**.

Reviewer 2: ```suggestion
1. On the page that opens, make your way to **Google**.
```

[straight from the spec](https://w3c.github.io/gamepad/#fig-visual-representation-of-a-standard-gamepad-layout:~:text=Figure%201%20Visual%20representation%20of%20a%20standard%20gamepad%20layout.),
shows the mapping and the arrangement of the buttons and axes on a generic gamepad.

<figure class="w-figure">
  <img src="gamepad.svg"
       alt="Schematic overview of the button and axes mappings of a common gamepad."
  >
  <figcaption class="w-figcaption">
    Visual representation of a standard gamepad layout
    (<a href="https://w3c.github.io/gamepad/#fig-visual-representation-of-a-standard-gamepad-layout:~:text=Figure%201%20Visual%20representation%20of%20a%20standard%20gamepad%20layout.">Source</a>).
  </figcaption>
</figure>

### Being notified when a gamepad gets connected

To learn when a gamepad is connected, listen for the `gamepadconnected`
event that triggers on the `window` object.
When the user connects a gamepad, which can either happen via USB or via Bluetooth,
a `GamepadEvent` is fired that has the gamepad's details in an aptly named `gamepad` property.
Below, you can see an example from an Xbox 360 controller that I had lying around
(yes, I am into retro gaming).

> ```js/3-12

Reviewer: I fee like the highlighting is not necessary at this point.

Author: Fair remark.

> ```js/3-12

Author: ```suggestion
```js
```

window.addEventListener('gamepadconnected', (event) => {
  console.log('✅ 🎮 A gamepad was connected:', event.gamepad);
  /*
    gamepad: Gamepad
    axes: (4) [0, 0, 0, 0]
    buttons: (17) [GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton]
    connected: true
    id: "Xbox 360 Controller (STANDARD GAMEPAD Vendor: 045e Product: 028e)"
    index: 0
    mapping: "standard"
    timestamp: 6563054.284999998
    vibrationActuator: GamepadHapticActuator {type: "dual-rumble"}
  */
});
```

### Being notified when a gamepad gets disconnected

Being notified of gamepad disconnects happens analogously to the way connections are detected.

> This time the app listens for the `gamepaddisconnect` event.

Reviewer: s/gamepaddisconnect/gamepaddisconnected/g

Author: Introduced during the code review. Thanks for the eagle eye!

> This time the app listens for the `gamepaddisconnect` event.

Author: ```suggestion
This time the app listens for the `gamepaddisconnected` event.
```

Note how in the example below `connected` is now `false`
when I unplug the Xbox 360 controller.

```js/6

> window.addEventListener('gamepadconnected', (event) => {

Reviewer: I guess it should be `'gamepaddisconnected'`.

Author: Yes, absolutely. Copy & paste…

> window.addEventListener('gamepadconnected', (event) => {

Author: ```suggestion
window.addEventListener('gamepaddisconnected', (event) => {
```

  console.log('❌ 🎮 A gamepad was disconnected:', event.gamepad);
  /*
    gamepad: Gamepad
    axes: (4) [0, 0, 0, 0]
    buttons: (17) [GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton, GamepadButton]
    connected: false
    id: "Xbox 360 Controller (STANDARD GAMEPAD Vendor: 045e Product: 028e)"
    index: 0
    mapping: "standard"
    timestamp: 6563054.284999998

>     vibrationActuator: null

Reviewer: No more `vibrationActuator`?

Author: I don't know why it's forgotten, it reports `null`.

  */
});
```

### The gamepad in your game loop

Getting a hold of a gamepad starts with a call to `navigator.getGamepads()`,
which returns a `GamepadList` object with `Gamepad` items.
The `GamepadList` object in Chrome *always* has a fixed length of four items.
If zero or less than four gamepads are connected, an item may just be `null`.
Always be sure to check all items of the `GamepadList` and be aware that
gamepads "remember" their slot and may not always be present at the first available slot.

```js
// When no gamepads are connected:

> navigator.getGamepads()

Reviewer: ```suggestion
navigator.getGamepads();

  motors, normalized to the range `0.0`–`1.0`.

```js
// This assumes a `Gamepad` as the value of the `gamepad` variable.
const vibrate = (gamepad, delay = 0, duration = 100, weak = 1.0, strong = 1.0) {
  if (!('vibrationActuator' in gamepad)) {
    return;
  }
  gamepad.vibrationActuator.playEffect('dual-rumble', {
    // Start delay in ms.
    startDelay: delay,
    // Duration is ms.
    duration: duration,
    // The magnitude of the weak actuator (between 0 and 1).
    weakMagnitude: weak,
    // The magnitude of the strong actuator (between 0 and 1).
    strongMagnitude: strong,
  });
};
```

### Integration with Permissions Policy

The Gamepad API spec defines a
[policy-controlled feature](https://w3c.github.io/webappsec-permissions-policy/)
identified by the string `"gamepad"`. Its default `allowlist` is `"self"`.
A document's permissions policy determines whether any content in that document is allowed
to access `navigator.getGamepads()`.
If disabled in any document, no content in the document will be allowed to use
`navigator.getGamepads()`, nor will the `gamepadconnected` and
`gamepaddisconnected` events fire.

> ```html/2

Reviewer: I'm not sure having lines highlighted helps. One line would be enough for me.

Author: Indeed, it's really short enough.

> ```html/2
> <iframe
>     src="index.html"
>     allow="gamepad"
> >
> </iframe>

Author: ```suggestion
```html
<iframe src="index.html" allow="gamepad"></iframe>
```

```

## Demo

A simple [gamepad tester demo](https://gamepad-demo.glitch.me/) is embedded below.
The source code is available [on Glitch](https://glitch.com/edit/#!/gamepad-demo).

> Try the demo by connecting a gamepad via USB or Bluetooth and pressing any of its buttons

Reviewer: The "vibrate" button doesn't work on Joy-Con L device. WAI?

Author: I *think* having read that the Joy-Cons internally have one heavy and one light motor. They can only shine when you hold them together as one gamepad, not separated.

Author: I changed the demo to use both strong and weak.

or moving any of its axis.

> <div class="glitch-embed-wrap" style="height: 1000px; width: 100%;">

Reviewer: I believe you can use something like:
```
{% Glitch { id: 'gamepad-demo', path: 'script.js', height: 480, allow: 'gamepad' } %}
```

Author: Done

> <div class="glitch-embed-wrap" style="height: 1000px; width: 100%;">
>   <iframe
>     src="https://glitch.com/embed/#!/embed/gamepad-demo?path=index.html&previewSize=100"
>     title="gamepad-demo on Glitch"
>     allow="gamepad"
>     style="height: 100%; width: 100%; border: 0;">
>   </iframe>
> </div>

Author: ```suggestion
{% Glitch { id: 'gamepad-demo', path: 'script.js', height: 1000, allow: 'gamepad' } %}
```
